examples/global_init/hpx_initialize.hpp

```diff
@@ -0,0 +1,107 @@
+
+#include <hpx/hpx.hpp>
+#include <hpx/hpx_init.hpp>
+
+///////////////////////////////////////////////////////////////////////////////
+// Store the command line arguments in global variables to make them available
+// to the startup code.
+
+#if defined(linux) || defined(__linux) || defined(__linux__)
+
+int __argc = 0;
+char** __argv = nullptr;
+
+void set_argc_argv(int argc, char* argv[], char* env[])
+{
+    __argc = argc;
+    __argv = argv;
+}
+
+__attribute__((section(".preinit_array")))
+    void (*set_global_argc_argv)(int, char*[], char*[]) = &set_argc_argv;
+
+#elif defined(__APPLE__)
+
+#include <crt_externs.h>
+
+inline int get_arraylen(char** argv)
+{
+    int count = 0;
+    if (nullptr != argv)
+    {
+        while(nullptr != argv[count])
+        ++count;
+    }
+    return count;
+}
+
+int __argc = get_arraylen(*_NSGetArgv());
+char** __argv = *_NSGetArgv();
+
+#endif
+
+///////////////////////////////////////////////////////////////////////////////
+// main function declaration
+int main(int argc, char* argv[]);
+
+///////////////////////////////////////////////////////////////////////////////
+// This class demonstrates how to initialize the hpx runtime system with main
+// running as an HPX thread.
+//
+struct manage_global_runtime {
+    manage_global_runtime() {
+
+        std::vector<std::string> const cfg = {
+            // allow for unknown command line options
+            "hpx.commandline.allow_unknown!=1",
+            // disable HPX' short options
+            "hpx.commandline.aliasing!=0"
+        };
+
+        using hpx::util::placeholders::_1;
+        using hpx::util::placeholders::_2;
+        hpx::util::function_nonser<int(int, char**)> start_function =
+            hpx::util::bind(&main, _1, _2);
+
+        // Start main on an HPX thread and then suspend this thread until
+        // hpx::finalize() is called
+        hpx::init(start_function, __argc, __argv, cfg, hpx::runtime_mode_console);
+    }
+
+    ~manage_global_runtime() {
+    // Destructor functions to go here
+    }
+};
+// On object declaration the hpx runtime system is initialized
+manage_global_runtime init;
+
+///////////////////////////////////////////////////////////////////////////////
+// This class is provides the necessary medium to de-initialize everything
+// and then safely exit the program
+//
+struct destruct {
+    destruct() {
+        // Exit the hpx runtime system
+        hpx::finalize();
+        // Call init destructor to de-initialize everything before exiting
+        init.~manage_global_runtime();
+
+        // Safely exit the program
+        std::exit(EXIT_SUCCESS);
+    }
+};
+
+//////////////////////////////////////////////////////////////////////////////
+// This implemented function calls the destruct object to start the exit
+// sequence
+namespace hpx {
+    void complete() {
+        // Finalize everything
+        destruct destroy;
+    }
+}
```

Comment:
Why is this necessary? Wouldn't invoking hpx::finalize() from the main function of the application trigger the destruction of the global object?

Reply:
@hkaiser I was not aware if hpx::finalize() trigger destruction of the global objects as well. So, I created an alternative way of hpx::complete() which would first destruct runtime system then de-initialize the initialization sequences that will be implemented in manage_global_runtime.

Reply:
`hpx::finalize()` informs the runtime that it is safe to exit. This should eventually cause the runtime to exit and the main application should terminate. At this point the system calls the destructors of all global objects (after the C-main function returned).

Reply:
ok, so calling hpx::finalize() in main itself should work properly. I'll make amendments to my code.

Reply:
Also, wasn't the goal of the whole example to show that the user does not need to invoke any special functionality in order to use HPX?

Reply:
What does this example do differently compared to the existing example `init_globally`?

Reply:
This example runs main directly as hpx thread without using hpx_main. This gives main function complete functionalities of an hpx thread i.e. futures and async can run without giving null id runtime error. In init_globally you could not get all functionalities of hpx thread in main, I received null id error for futures and async. This is because main is registered as hpx thread, so only a few components can be used.
Furthermore, I felt that this implementation is better since it completely drops the idea of hpx_main, which was the key aspect of my GSoC project (i.e. running things directly from main without going through a special function).